storage/src/cache/mod.rs

```diff
@@ -410,6 +422,42 @@ impl<'a, 'b> ChunkDecompressState<'a, 'b> {
         }
     }
 
+    fn decompress_batch(&mut self, meta: &Arc<BlobCompressionContextInfo>) -> Result<()> {
+        let ctx = meta
+            .get_batch_context(self.batch_idx)
+            .ok_or_else(|| einval!("failed to get Batch context for chunk"))?;
+        let c_offset = ctx.compressed_offset();
+        let c_size = ctx.compressed_size() as u64;
+        let d_size = ctx.uncompressed_batch_size() as u64;
+        if c_offset < self.blob_offset
+            || c_offset.checked_add(c_size).is_none()
+            || c_offset + c_size > self.blob_offset + self.c_buf.len() as u64
+            || d_size > RAFS_MAX_CHUNK_SIZE
+        {
+            let msg = format!(
+                "invalid chunk: z_offset 0x{:x}, z_size 0x{:x}, c_offset 0x{:x}, c_size 0x{:x}, d_size 0x{:x}",
+                self.blob_offset,
+                self.c_buf.len(),
+                c_offset,
+                c_size,
+                d_size
+            );
+            return Err(einval!(msg));
+        }
+
+        let c_offset = (c_offset - self.blob_offset) as usize;
+        let input = &self.c_buf[c_offset..c_offset + c_size as usize];
+        let mut output = alloc_buf(d_size as usize);
+
+        self.cache
+            .decompress_chunk_data(input, &mut output, c_size != d_size)?;
+        assert!(output.len() == d_size as usize);
```

Comment:
Please return error instead of assert, otherwise it may cause DoS attack by injecting bad image data.

Reply:
This assertion depends data from user input, and we can't trust user here.